discovery/gossiper.go

```diff
@@ -2359,70 +2327,270 @@ func (d *AuthenticatedGossiper) fetchNodeAnn(
+		peerPubKey)
+
+	// Once we detect the peer has reconnected, we'll also request a
+	// notification for when they disconnect. We'll use this to make sure
+	// they haven't disconnected (in the case of a flappy peer, etc.) by the
+	// time we attempt to send them the pending messages.
+	log.Debugf("Requesting offline notification for peer=%x", peerPubKey)
+
+	offlineChan := d.cfg.NotifyWhenOffline(peerPubKey)
+
+	pendingMsgs, err := d.cfg.MessageStore.MessagesForPeer(peerPubKey)
+	if err != nil {
+		log.Errorf("Unable to retrieve pending messages for "+
+			"peer %x: %v", peerPubKey, err)
+		goto waitUntilOnline
+	}
+
+	// With the peer online, we can now proceed to send our pending messages
+	// for them.
+	allMsgsStale := true
+	for _, msg := range pendingMsgs {
+		// Retrieve the short channel ID for which this message applies
+		// for logging purposes. The error can be ignored as the store
+		// can only contain messages which have a ShortChannelID field.
+		shortChanID, _ := msgShortChanID(msg)
+
+		// Make sure that the peer has not disconnected by this point.
+		// If they have, we'll wait until they reconnect.
+		select {
+		case <-offlineChan:
+			goto waitUntilOnline
+		case <-d.quit:
+			return
+		default:
+		}
+
+		if err := peer.SendMessage(false, msg); err != nil {
+			log.Errorf("Unable to send %v message for channel=%v "+
+				"to %x: %v", msg.MsgType(), shortChanID,
+				peerPubKey, err)
+
+			// We'll mark allMsgsStale false to ensure this message
+			// gets sent at least once before being removed.
+			allMsgsStale = false
+
+			continue
+		}
+
+		log.Debugf("Successfully sent %v message for channel=%v with "+
+			"peer=%x upon reconnection", msg.MsgType(), shortChanID,
+			peerPubKey)
+
+		// Now that the message has at least been sent once, we can
+		// check whether it's stale. This guarantees that
+		// AnnounceSignatures are sent at least once if we happen to
+		// already have signatures for both parties.
+		if d.isStaleMsg(msg) {
+			log.Debugf("Removing stale %v message for channel=%v "+
+				"with peer=%x", msg.MsgType(), shortChanID,
+				peerPubKey)
+
+			err := d.cfg.MessageStore.DeleteMessage(msg, peerPubKey)
+			if err != nil {
+				log.Errorf("Unable to remove stale %v message "+
+					"for channel=%v with peer %x: %v",
+					msg.MsgType(), shortChanID, peerPubKey,
+					err)
 			}
 
-			log.Errorf("unable to send AnnounceSignatures message "+
-				"to peer(%x): %v. Will retry when online.",
-				remotePeer.SerializeCompressed(), err)
+			continue
+		}
 
-			peerChan := make(chan lnpeer.Peer, 1)
-			d.cfg.NotifyWhenOnline(remotePeer, peerChan)
+		allMsgsStale = false
+	}
 
-			select {
-			case <-peerChan:
-				// Retry sending.
-				log.Infof("Peer %x reconnected. Retry sending"+
-					" AnnounceSignatures.",
-					remotePeer.SerializeCompressed())
+	// If all of our messages were stale, then there's no need for this
+	// handler to continue running, so we can exit now.
+	if allMsgsStale {
```

Comment:
can be simplified to just get the current set of messages from the db, and checking whether it is empty.

Reply:
Fixed.